src/rust/catloop/socket.rs

```diff
@@ -0,0 +1,670 @@
+            Ok((new_qd, remote))
+        };
+
+        match result {
+            Ok((new_qd, remote)) => {
+                let mut socket: RefMut<Socket> = socket.borrow_mut();
+                socket.state.commit();
+                socket.catmem_qd = Some(new_qd);
+                socket.remote = Some(remote);
+                Ok(())
+            },
+            Err(e) => {
+                socket.borrow_mut().state.rollback();
+                Err(e)
+            },
+        }
+    }
+
+    /// Closes this socket.
+    pub fn close(&mut self) -> Result<(), Fail> {
+        self.state.prepare(SocketOp::Close)?;
+        self.state.commit();
+        self.state.prepare(SocketOp::Closed)?;
+        if let Some(qd) = self.catmem_qd {
+            match self.catmem.borrow_mut().close(qd) {
+                Ok(()) => {
+                    self.state.commit();
+                    return Ok(());
+                },
+                Err(e) => {
+                    self.state.rollback();
+                    return Err(e);
+                },
+            }
+        };
+        self.state.commit();
+        Ok(())
+    }
+
+    /// Asynchronously closes this socket by allocating a coroutine.
+    pub fn async_close<F>(&mut self, coroutine: F, yielder: Yielder) -> Result<(QToken, bool), Fail>
+    where
+        F: FnOnce(Yielder) -> Result<TaskHandle, Fail>,
+    {
+        self.state.prepare(SocketOp::Close)?;
+        if let Some(qd) = self.catmem_qd {
+            match self.catmem.borrow().async_close(qd) {
+                Ok(qt) => {
+                    self.state.commit();
+                    Ok((qt, false))
+                },
+                Err(e) => {
+                    self.state.rollback();
+                    Err(e)
+                },
+            }
+        } else {
+            Ok((
+                self.do_generic_sync_control_path_call(coroutine, yielder)?
+                    .get_task_id()
+                    .into(),
+                true,
+            ))
+        }
+    }
+
+    pub fn do_close(socket: Rc<RefCell<Self>>) -> Result<(), Fail> {
+        let mut socket: RefMut<Self> = socket.borrow_mut();
+        socket.state.prepare(SocketOp::Closed)?;
+        socket.state.commit();
+        Ok(())
+    }
+
+    /// Pushes to the underlying Catmem queue that is the transport for this socket.
+    pub fn push(&self, sga: &demi_sgarray_t) -> Result<QToken, Fail> {
+        match self.catmem_qd {
+            Some(qd) => self.catmem.borrow().push(qd, sga),
+            None => Err(Fail::new(libc::EINVAL, "pipe not connected")),
+        }
+    }
+
+    /// Pop from the underlying Catmem queue that is the transport for this socket.
+    pub fn pop(&self, size: Option<usize>) -> Result<QToken, Fail> {
+        match self.catmem_qd {
+            Some(qd) => self.catmem.borrow().pop(qd, size),
+            None => Err(Fail::new(libc::EINVAL, "pipe not connected")),
```

Comment:
Is this the same error that we return for the other LibOSes? Shouldn't it be `ENOTCONN`?